docs/howtos/DataPlane Generation - DPG/09versioning.mdx

```diff
@@ -0,0 +1,406 @@
+
+stable_model = models.StableModel(stable_feature="present")
+print(stable_model)
+preview_client.stable_functionality(stable_model) # call goes through
+
+preview_model = models.PreviewModel(preview_functionality="present")
+# the model is generated as part of the api surface
+
+preview_client.preview_functionality(preview_model) # call goes through
+```
+
+</TabItem>
+<TabItem value="csharp" label="CSharp" >
+
+```csharp
+
+```
+
+</TabItem>
+<TabItem value="typescript" label="Typescript" >
+
+```typescript
+
+```
+
+</TabItem>
+<TabItem value="java" label="Java" >
+
+```java
+
+```
+
+</TabItem>
+<TabItem value="go" label="Go" >
+
+```go
+
+```
+
+</TabItem>
+</Tabs>
+
+### Override to return all
+
+You can also override the signature to return the combined api surface area of all of the separate api versions. Different languages have different support for versioning validation
+
+<Tabs>
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+import "@typespec/versioning";
+
+using TypeSpec.Versioning;
+
+@versioned(My.Service.Versions)
+@service
+namespace My.Service;
+
+enum Versions {
+  v2023_11_01_preview: "2023-11-01-preview",
+  v2023_11_01: "2023-11-01",
+}
+
+model PreviewModel {
+  betaFeature: string;
+}
+
+model StableModel {
+  stableFeature: string;
+}
+
+@added(Versions.v2023_11_01_preview)
+@removed(Versions.v2023_11_01)
+op previewFunctionality(...PreviewModel): void;
+
+op stableFunctionality(...StableModel): void;
+```
+
+</TabItem>
+
+<TabItem value="tspconfig.yaml" label="tspconfig.yaml" default>
+
+```yaml
+---
+options:
+  "@azure-tools/typespec-python":
+    api-version: "all"
+  "@azure-tools/typespec-csharp":
+    api-version: "all"
+  "@azure-tools/typespec-ts":
+    api-version: "all"
+  "@azure-tools/typespec-java":
+    api-version: "all"
+  "@azure-tools/typespec-go":
+    api-version: "all"
+```
+
+</TabItem>
+
+<TabItem value="python" label="Python">
+
+```python
+import pytest
+from my.service import MyServiceClient, models
+
+client = MyServiceClient(endpoint=..., credential=...)
+# client's api_version will be "2023-11-01"
+
+stable_model = models.StableModel(stable_feature="present")
+print(stable_model)
+client.stable_functionality(stable_model) # call goes through
+
+preview_model = models.PreviewModel(preview_functionality="present")
+# the model is generated as part of the api surface
+
+with pytest.expect(ValueError) as ex:
+  client.preview_functionality(preview_model)
+assert "preview_functionality is not available in api version 2023-11-01" in str(ex)
+
+preview_client = MyServiceClient(endpoint=..., credential=..., api_version="2023-11-01-preview")
+
+preview_client.preview_functionality(preview_model) # call goes through
+```
+
+</TabItem>
+<TabItem value="csharp" label="CSharp" >
+
+```csharp
+
+```
+
+</TabItem>
+<TabItem value="typescript" label="Typescript" >
+
+```typescript
+
+```
+
+</TabItem>
+<TabItem value="java" label="Java" >
+
+```java
+
```

Comment:
Discussion:

Even it is `"all"`, I don't think Java would ever want to generate a API not in stable api-version to a SDK in stable api-version.
Java likely won't generate a removed API (or property) in any case, especially for GA version.

Reply:
This discussion may be a bit on the real case, instead of a sample

Let's assume the versions in tsp be

```ts
enum Versions {
  v2023_11_01_preview: "2023-11-01-preview",
  v2023_11_01: "2023-11-01",
  v2023_12_01_preview: "2023-12-01-preview",
  v2023_12_01: "2023-12-01",
}
```

When release SDK for "v2023_12_01", typically it only want to generate for version
```
v2023_11_01
v2023_12_01
```

If any API / property only exists in preview version, SDK won't want to include them.

---

In data-plane, this may not be an issue from typespec source, as reviewer would suggest service to remove any preview prior to stable, so their typespec would only contain
```ts
enum Versions {
  v2023_11_01: "2023-11-01",
  v2023_12_01: "2023-12-01",
}
```
at this stage.

If there is future preview, it would be append to the list. Then the next stable would clear these previews.

However, this may not be the case for management.

Reply:
For `all`, I don't think we have a clear decision for how to deal with version resiliency. I suggest removing this part for now.

Reply:
`all` just means "no projection". Meaning you get access to the full list of things, and you have access to `api_versions` in each object to decide your own logic, if you want to.
This is literally what we have today, we all receive "all", and we made it work.
I'm ok to add a new conf option like "stable_only", but can someone explains to me how it would be different from generating the latest?

Reply:
specifying `2023-12-01` will not get any removal or say breaking changes caused by previous stable version because of projection. use `all` will not do projection, so all info will be there.

Reply:
I am OK if this is just a case or demonstration.

But personally I would avoid advice service set this "all". At present, I don't see a convincing case that service really need to set "all" to handle their SDK.